modules/imgcodecs/src/grfmt_spng.cpp

```diff
@@ -0,0 +1,520 @@
+// indirect, incidental, special, exemplary, or consequential damages
+// (including, but not limited to, procurement of substitute goods or services;
+// loss of use, data, or profits; or business interruption) however caused
+// and on any theory of liability, whether in contract, strict liability,
+// or tort (including negligence or otherwise) arising in any way out of
+// the use of this software, even if advised of the possibility of such damage.
+//
+//M*/
+
+#include "precomp.hpp"
+
+#ifdef HAVE_SPNG
+
+/****************************************************************************************\
+    This part of the file implements PNG codec on base of libspng library,
+    in particular, this code is based on example.c from libspng
+    (see 3rdparty/libspng/LICENSE for copyright notice)
+\****************************************************************************************/
+
+#ifndef _LFS64_LARGEFILE
+#  define _LFS64_LARGEFILE 0
+#endif
+#ifndef _FILE_OFFSET_BITS
+#  define _FILE_OFFSET_BITS 0
+#endif
+
+#include <spng.h>
+#include <zlib.h>
+
+#include "grfmt_spng.hpp"
+
+namespace cv
+{
+
+/////////////////////// PngDecoder ///////////////////
+
+    SPngDecoder::SPngDecoder()
+    {
+        m_signature = "\x89\x50\x4e\x47\xd\xa\x1a\xa";
+        m_color_type = 0;
+        m_ctx = 0;
+        m_f = 0;
+        m_buf_supported = true;
+        m_buf_pos = 0;
+        m_bit_depth = 0;
+    }
+
+
+    SPngDecoder::~SPngDecoder()
+    {
+        close();
+    }
+
+    ImageDecoder SPngDecoder::newDecoder() const
+    {
+        return makePtr<SPngDecoder>();
+    }
+
+    void  SPngDecoder::close()
+    {
+        if( m_f )
+        {
+            fclose( m_f );
+            m_f = 0;
+        }
+
+        if( m_ctx ) {
+            struct spng_ctx *ctx = (struct spng_ctx*)m_ctx;
+            spng_ctx_free(ctx);
+            m_ctx = 0;
+        }
+    }
+
+
+    int  SPngDecoder::readDataFromBuf( void* sp_ctx, void *user, void* dst, size_t size )
+    {
+        /*
+         * typedef int spng_read_fn(spng_ctx *ctx, void *user, void *dest, size_t length)
+         *   Type definition for callback passed to spng_set_png_stream() for decoders.
+         *   A read callback function should copy length bytes to dest and return 0 or SPNG_IO_EOF/SPNG_IO_ERROR on error.
+         */
+        SPngDecoder* decoder = (SPngDecoder*)(user);
+        CV_Assert( decoder );
+
+        const Mat& buf = decoder->m_buf;
+        if( decoder->m_buf_pos + size > buf.cols*buf.rows*buf.elemSize() )
+        {
+            return SPNG_IO_ERROR;
+        }
+        memcpy( dst, decoder->m_buf.ptr() + decoder->m_buf_pos, size );
+        decoder->m_buf_pos += size;
+
+        return 0;
+    }
+
+    bool  SPngDecoder::readHeader()
+    {
+        volatile bool result = false;
+        close();
+
+        spng_ctx *ctx = spng_ctx_new(0);
+
+        if(!ctx) {
+            spng_ctx_free(ctx);
```

Comment:
Please check if `spng_ctx_free` with NULL is expected behaviour.

Reply:
```cpp
void spng_ctx_free(spng_ctx *ctx)
{
    if(ctx == NULL) return;
```

It is expected behavior but it seems unnecessary. For the future releases of spng, it might be useful to keep this line.